src/VisualStudio/Core/Def/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs

```diff
@@ -207,79 +188,155 @@ public int OnBeforeLastDocumentUnlock(uint docCookie, uint dwRDTLockType, uint d
+            if (languageInformation == null)
             {
-                // See if we should push to this to the metadata-to-source workspace instead.
-                if (_runningDocumentTable.IsDocumentInitialized(docCookie))
+                // We can never put this document in a workspace, so just bail
+                return;
+            }
+
+            if (_runningDocumentTable.IsDocumentInitialized(docCookie))
+            {
+                var vsTextBuffer = (IVsTextBuffer)_runningDocumentTable.GetDocumentData(docCookie);
+                var textBuffer = _editorAdaptersFactoryService.GetDocumentBuffer(vsTextBuffer);
+
+                // As long as the buffer is initialized, then we should see if we should attach
+                if (textBuffer != null)
                 {
-                    var vsTextBuffer = (IVsTextBuffer)_runningDocumentTable.GetDocumentData(docCookie);
-                    var textBuffer = _editorAdaptersFactoryService.GetDocumentBuffer(vsTextBuffer);
+                    var registration = Workspace.GetWorkspaceRegistration(textBuffer.AsTextContainer());
 
-                    if (_fileTrackingMetadataAsSourceService.TryAddDocumentToWorkspace(moniker, textBuffer))
+                    registration.WorkspaceChanged += Registration_WorkspaceChanged;
+                    _docCookieToWorkspaceRegistration = _docCookieToWorkspaceRegistration.Add(docCookie, registration);
+
+                    if (!IsClaimedByAnotherWorkspace(registration))
                     {
-                        // We already added it, so we will keep it excluded from the misc files workspace
-                        return;
+                        AttachToDocument(docCookie, moniker);
                     }
                 }
+            }
+        }
 
-                var parseOptions = languageInformation.ParseOptions;
+        private void Registration_WorkspaceChanged(object sender, EventArgs e)
+        {
+            var workspaceRegistration = (WorkspaceRegistration)sender;
+            uint docCookie;
+            Contract.ThrowIfFalse(_docCookieToWorkspaceRegistration.TryGetKey(workspaceRegistration, out docCookie));
+            var moniker = _runningDocumentTable.GetDocumentMoniker(docCookie);
 
-                if (Path.GetExtension(moniker) == languageInformation.ScriptExtension)
+            if (workspaceRegistration.Workspace == null)
+            {
+                // We should now claim this
+                AttachToDocument(docCookie, moniker);
+            }
+            else if (IsClaimedByAnotherWorkspace(workspaceRegistration))
+            {
+                // It's now claimed by another workspace, so we should unclaim it
+                if (_docCookiesToHostProject.ContainsKey(docCookie))
                 {
-                    parseOptions = parseOptions.WithKind(SourceCodeKind.Script);
+                    DetachFromDocument(docCookie, moniker);
                 }
+            }
```

Comment:
Yes, one case is we're moving between misc files and metadata as source....which doesn't really make sense.